crates/api/src/v1/monitor.rs

```diff
@@ -0,0 +1,236 @@
+//! Types relating to the monitor API.
+
+use crate::Status;
+use serde::{de::Unexpected, Deserialize, Serialize, Serializer};
+use std::borrow::Cow;
+use thiserror::Error;
+use warg_crypto::hash::AnyHash;
+use warg_crypto::signing;
+use warg_protocol::registry::RegistryLen;
+
+/// Represents checkpoint verification response in one of the following states:
+#[derive(Serialize, Deserialize)]
+#[serde(tag = "state", rename_all = "camelCase")]
+#[allow(clippy::large_enum_variant)]
+pub enum CheckpointVerificationResponse {
+    /// The checkpoint is verified.
+    #[serde(rename_all = "camelCase")]
+    Verified,
+    /// The checkpoint is verified but has not verified the signature.
+    #[serde(rename_all = "camelCase")]
+    VerifiedButSignatureUnverified,
+    /// The checkpoint is unverified but should retry verification.
+    #[serde(rename_all = "camelCase")]
+    Retry {
+        /// Instructs to retry but wait the specified number of seconds.
+        wait_seconds: u16,
+    },
+    /// The checkpoint is unverified and should not retry.
+    #[serde(rename_all = "camelCase")]
+    Unverified,
+}
+
+/// Represents a monitor API error.
+#[non_exhaustive]
+#[derive(Debug, Error)]
+pub enum MonitorError {
+    /// The provided checkpoint log length is not known to the registry.
+    #[error("checkpoint log length `{0}` is not known to the registry")]
+    CheckpointNotFound(RegistryLen),
+    /// The provided checkpoint log length is greater than the most recent checkpoint. The monitor
+    /// is expected to only return this after it is certain that it has the most recent checkpoint.
+    #[error("checkpoint log length `{0}` is invalid and greater than the most recent checkpoint")]
+    CheckpointInvalidLogLength(RegistryLen),
```

Comment:
I'm not sure that this error makes sense as written. In a distributed system you never really know that you "have the most recent" of something...

Reply:
This is attempting to handle the case where the requested checkpoint is actually in the future. The thinking is that a monitor would first respond with `retry` and check back. And the monitor would check that it has the most recent checkpoint, and then on the next `retry` be able to answer definitely that it is not a valid `logLength`.

Let's say someone wanted to circumvent the verification process and consistently created invalid checkpoints in the future. Without some mechanism like this, the best the monitor could do is say, "I don't know". This gives the mechanism to actually return "not valid".